plenum/server/replica.py

```diff
@@ -819,27 +824,73 @@ def trackBatches(self, pp: PrePrepare, prevStateRootHash):
         self.batches[(pp.viewNo, pp.ppSeqNo)] = [pp.ledgerId, pp.discarded,
                                                  pp.ppTime, prevStateRootHash, len(pp.reqIdr)]
 
-    def send3PCBatch(self):
-        r = 0
-        for lid, q in self.requestQueues.items():
-            # TODO: make the condition more apparent
-            if len(q) >= self.config.Max3PCBatchSize or \
-                    (self.lastBatchCreated + self.config.Max3PCBatchWait < time.perf_counter() and
-                     len(q) > 0):
-                oldStateRootHash = self.stateRootHash(lid, to_str=False)
-                ppReq = self.create3PCBatch(lid)
-                if ppReq is None:
-                    continue
-                self.sendPrePrepare(ppReq)
-                if not self.isMaster:
-                    self.node.last_sent_pp_store_helper.store_last_sent_pp_seq_no(
-                        self.instId, ppReq.ppSeqNo)
-                self.trackBatches(ppReq, oldStateRootHash)
-                r += 1
-
-        if r > 0:
-            self.lastBatchCreated = time.perf_counter()
-        return r
+    def send_3pc_batch(self):
+        sent_batches = set()
+
+        # 1. send 3PC batches with requests for every ledger
+        self._send_3pc_batches_for_ledgers(sent_batches)
+
+        # 2. for every ledger we havne't just sent a 3PC batch check if it's not fresh enough,
+        # and send an empty 3PC batch to update the state if needed
+        self._send_3pc_freshness_batch(sent_batches)
+
+        # 3. update ts of last sent 3PC batch
+        sent_batches.discard(None)
+        if len(sent_batches) > 0:
+            self.lastBatchCreated = self.get_current_time()
+
+        return len(sent_batches)
+
+    def _send_3pc_batches_for_ledgers(self, sent_batches):
+        for ledger_id, q in self.requestQueues.items():
+            if len(q) == 0:
+                continue
+
+            queue_full = len(q) >= self.config.Max3PCBatchSize
+            timeout = self.lastBatchCreated + self.config.Max3PCBatchWait < self.get_current_time()
+            if not queue_full and not timeout:
+                continue
+
+            sent_batches.add(
```

Comment:
Probably changing this to
```python
if queue_full or timeout:
    send_batches.add(...)
```
make it a bit more readable.
Another option is to do just one check per continue, like:
```
if len(q) < max_batch_size:
   continue
if self.get_current_time() < self.lastBatchCreated + max_batch_wait:
   continue
send_batches.add(...)
```
This all is pretty subjective though.

Reply:
The second option (one check per continue) will not work,
We can exit because the queue is not full (len(q) < max_batch_size) while we still need to send the batch because of timeout.

In general, this is quite subjective, but I don't like nested if-else and loops.
So, I prefer the pattern
```
if not check1:
    return/continue
if not check2:
    return/continue
.....
do_work

```